README.md
Swift metadata sync
===================

Synchronize Swift metadata with an Elasticsearch index. The tool can be used
through the `swift-metadata-sync` binary. The synchronization daemon needs to
run on a Swift node that hosts a container database.

The daemon will attempt to ensure correct mappings for the fields it knows about
(e.g. x-timestamp, last-modified, content-length, etag). However, if an
incorrect mapping has already been created, the metadata will still be
propagated, but may not be searchable (e.g. a range query against dates would
not return expected results).

Usage
-----

The bare minimum required is a configuration file. It must include a pointer to
the Swift drives (`devices`), the status directory to use (`status_dir`), number
of items to process at a time (`items_chunk`), and an array of container mappings.

Here is a sample configuration file:

	{
		"devices": "/srv/node",
		"status_dir": "/tmp",
		"items_chunk": 1000,
		"containers": [
			{
				"account": "AUTH_swift",
				"container": "swift",
				"es_hosts": "192.168.22.1",
				"index": "stuff"
			}
		]
	}


For each Swift Account/Container, an elasticsearch cluster (`es_hosts`) and
index (`index`) must be specified. The hosts argument accepts multiple,
comma-separated entries to specify numerous servers.

If an index is changed and a re-index is desired, changing a container mapping's
`index` value will restart indexing from the first object in that container.

Design
------

The daemon walks the Swift container database present on the node. The database
rows contain the names of the objects and their status (notably last modified
date and whether the object has been deleted). When an object's metadata is
mutated, as long as fast-POST is enabled, a new row will be inserted in the
database and the prior entry removed. This allows the daemon to continually walk
the databases rows forward.

The advantage of this approach is that we never have to scan the entire
database. The daemons must run on each of the container nodes to ensure that
metadata is propagated, even if some of the container nodes fail.

There is no coordination mechanism between the processes walking the database,
but each one attempts to only process a fraction of the entries and then
verifies that all entries have been propagated.

After a failure, a daemon can be safely replaced with another node. It will only
be doing bulk queries against Elasticsearch to verify that the changes it
expects to have observed have been made. Once it catches up, it will resume
operation from where the failure occurred.

The daemons also record the database ID. If a drive fails and the database has
to replicated from another Swift node, the daemon will also restart from the
beginning. This ensures correctness, but means that updates may not propagate as
quickly after drive failures.

Testing it out
--------------

You can build a docker container with a Swift all-in-one and elasticsearch to
try out metadata search. The container is defined in `test/container`. To build
it, run: `docker build -t metadata-sync test/container` (this will tag the
docker image as `metadata-sync`). Once the container is built, you can launch it
as follows (assuming you're in the root directory of the swift-metadata-sync
repository):
``docker run -P -d -v `pwd`:/swift-metadata-sync metadata-sync``.

This will create a container running in the background (`-d`) with three ports
exposed (`-P`), which maps the code tree into the container at
`/swift-metadata-sync`. `swift` is listening on port 8080 inside the container
and elasticsearch is on 9200. To check the port mappings, use:
`docker port <container-name>`.

Once the container is running, you can use the Swift cluster as expected. The
default mappings are configured in `test/container/swift-metadata-sync.json`. If
you create the `es-test` container and an index named `es-test`, you should see
the objects' metadata appear in elasticsearch.


CHANGELOG.md
## 0.0.11 (2018-06-08)

Features:

    - Allow for parsing JSON from metadata values.
    - Operators can define the Elasticsearch pipeline to be used when indexing
      documents.

## 0.0.10 (2017-09-11)

Bug fixes:

    - Fix an issue where objects with unicode names cannot be indexed if
      simple-json is not installed (as the UTF-8 encoded strings may be
      attempted to be encoded one more time).
    - Handles a missing type error gracefully, which may occur if the index was
      created without a mapping. In that case, the "object" document type is
      created with the default mappings.

## 0.0.9 (2017-07-13)

Bug fixes:

    - change the constructor to work with the changes to the Container Crawler.

## 0.0.8 (2017-07-12)

Bug fixes:

    - properly handle unicode characters in object names.

## 0.0.7 (2017-06-12)

Improvement:

    - do not rely on the deprecated "found" field in DELETE responses.
    - use "text" or "keyword" fields when creating mappings for Swift objects
      (as opposed to "string"). We will now check the Elasticsearch version and
      only use "string" with Elasticsearch servers < 5.x.
    - bump the Elasticsearch library requirement to 5.x.

Bug fixes:

    - change the document ID to be SHA256 of account, container, and object. The
      string is concatenated with the "/" separator. This will ensures that we
      can work with long object names and Elasticsearch version 5.0 or newer,
      which reject IDs longer than 512 characters.

## 0.0.6 (2017-05-05)

Improvements:

    - change the handle() to conform to the new API in ContainerCrawler.

## 0.0.5 (2017-04-24)

Bug fixes:

    - exit with an error code of 0 and a message if the configuration file does
      not exist. This may happen on a fresh installation of the daemon.

## 0.0.4 (2017-02-28)

Bug fixes:

    - use long for the object size; otherwise, we are limited to reporting
      objects only up to 2GB.

## 0.0.3 (2017-02-06)

Bug fixes:

    - fix the conversion of last-modified to a proper Elasticsearch date format
      (milliseconds from epoch, as opposed to the mistaken use of seconds).

Improvements:

    - better error messages on failure to index (includes the "reason" from
      Elasticsearch).


test/container/launch.sh
#!/bin/bash

set -e

# Make sure all of the .pid files are removed -- services will not start
# otherwise
find /var/lib/ -name *.pid -delete
find /var/run/ -name *.pid -delete

# Copied from the docker swift container. Unfortunately, there is no way to
# plugin an additional invocation to start swift-s3-sync, so we had to do this.
/usr/sbin/service rsyslog start
/usr/sbin/service rsync start
/usr/sbin/service memcached start
# set up storage
mkdir -p /swift/nodes/1 /swift/nodes/2 /swift/nodes/3 /swift/nodes/4

for i in `seq 1 4`; do
    if [ ! -e "/srv/$i" ]; then
        ln -s /swift/nodes/$i /srv/$i
    fi
done
mkdir -p /srv/1/node/sdb1 /srv/2/node/sdb2 /srv/3/node/sdb3 /srv/4/node/sdb4 \
    /var/run/swift
/usr/bin/sudo /bin/chown -R swift:swift /swift/nodes /etc/swift /srv/1 /srv/2 \
    /srv/3 /srv/4 /var/run/swift
/usr/bin/sudo -u swift /swift/bin/remakerings

/usr/bin/sudo -u swift /swift/bin/startmain

PYTHONPATH=/opt/ss/lib/python2.7/dist-packages:/swift-metadata-sync \
    python -m swift_metadata_sync --log-level debug \
    --config /swift-metadata-sync/test/container/swift-metadata-sync.conf &

/usr/bin/sudo -u elastic /bin/bash /elasticsearch-5.5.2/bin/elasticsearch \
    2>&1 > /var/log/elasticsearch.log &

/usr/local/bin/supervisord -n -c /etc/supervisord.conf


swift_metadata_sync/__main__.py
import argparse
import json
import logging
import os
import traceback

from container_crawler import ContainerCrawler
from .metadata_sync import MetadataSync


def setup_logger(console=False, log_file=None, level='INFO'):
    logger = logging.getLogger('swift-metadata-sync')
    logger.setLevel(level)
    formatter = logging.Formatter(
        '[%(asctime)s] %(name)s [%(levelname)s]: %(message)s')
    if console:
        handler = logging.StreamHandler()
    elif log_file:
        handler = logging.handlers.RotatingFileHandler(log_file,
                                                       maxBytes=100 * 2**20,
                                                       backupCount=5)
    else:
        raise RuntimeError('log file must be set')
    handler.setFormatter(formatter)
    logger.addHandler(handler)


def load_config(conf_file):
    with open(conf_file, 'r') as f:
        return json.load(f)


def parse_args():
    parser = argparse.ArgumentParser(
        description='Swift metadata synchronization daemon')
    parser.add_argument('--config', metavar='conf', type=str, required=True,
                        help='path to the configuration file')
    parser.add_argument('--once', action='store_true',
                        help='run once')
    parser.add_argument('--log-level', metavar='level', type=str,
                        default='info',
                        choices=['debug', 'info', 'warning', 'error'],
                        help='logging level; defaults to info')
    parser.add_argument('--console', action='store_true',
                        help='log messages to console')
    return parser.parse_args()


def main():
    args = parse_args()
    if not os.path.exists(args.config):
        print 'Configuration file does not exist'
        exit(0)

    conf = load_config(args.config)
    setup_logger(console=args.console, level=args.log_level.upper(),
                 log_file=conf.get('log_file'))

    logger = logging.getLogger('swift-metadata-sync')
    logger.info('Starting Swift Metadata Sync')
    try:
        conf['bulk_process'] = True
        crawler = ContainerCrawler(conf, MetadataSync, logger)
        if args.once:
            crawler.run_once()
        else:
            crawler.run_always()
    except Exception as e:
        logger.error("Metadata Sync failed: %s" % repr(e))
        logger.error(traceback.format_exc(e))
        exit(1)


if __name__ == '__main__':
    main()


swift_metadata_sync/metadata_sync.py
from distutils.version import StrictVersion
import elasticsearch
import elasticsearch.helpers
import email.utils
import hashlib
import json
import logging
import os
import os.path

from swift.common.utils import decode_timestamps
from container_crawler.base_sync import BaseSync


class MetadataSync(BaseSync):
    DOC_TYPE = 'object'
    DOC_MAPPING = {
        "content-length": {"type": "long"},
        "content-type": {"type": "string"},
        "etag": {"type": "string", "index": "not_analyzed"},
        "last-modified": {"type": "date"},
        "x-object-manifest": {"type": "string"},
        "x-static-large-object": {"type": "boolean"},
        "x-swift-container": {"type": "string"},
        "x-swift-account": {"type": "string"},
        "x-swift-object": {"type": "string"},
        "x-timestamp": {"type": "date"},
        "x-trans-id": {"type": "string", "index": "not_analyzed"}
    }
    USER_META_PREFIX = 'x-object-meta-'

    def __init__(self, status_dir, settings, per_account=False):
        super(MetadataSync, self).__init__(status_dir, settings, per_account)

        self.logger = logging.getLogger('swift-metadata-sync')
        es_hosts = settings['es_hosts']
        self._es_conn = elasticsearch.Elasticsearch(es_hosts)
        self._server_version = StrictVersion(
            self._es_conn.info()['version']['number'])
        self._index = settings['index']
        self._parse_json = settings.get('parse_json', False)
        self._pipeline = settings.get('pipeline')
        self._verify_mapping()

    def get_last_row(self, db_id):
        if not os.path.exists(self._status_file):
            return 0
        with open(self._status_file) as f:
            try:
                status = json.load(f)
                entry = status.get(db_id, None)
                if not entry:
                    return 0
                if entry['index'] == self._index:
                    return entry['last_row']
                else:
                    return 0
            except ValueError:
                return 0
        return 0

    def save_last_row(self, row_id, db_id):
        if not os.path.exists(self._status_account_dir):
            os.mkdir(self._status_account_dir)
        if not os.path.exists(self._status_file):
            with open(self._status_file, 'w') as f:
                json.dump({db_id: dict(last_row=row_id,
                                       index=self._index)}, f)
                return

        with open(self._status_file, 'r+') as f:
            try:
                status = json.load(f)
            except ValueError:
                status = {}
            status[db_id] = dict(last_row=row_id, index=self._index)
            f.seek(0)
            json.dump(status, f)
            f.truncate()
            return

    def handle(self, rows, internal_client):
        self.logger.debug("Handling rows: %s" % repr(rows))
        if not rows:
            return []
        errors = []

        bulk_delete_ops = []
        mget_map = {}
        for row in rows:
            if row['deleted']:
                bulk_delete_ops.append({'_op_type': 'delete',
                                        '_id': self._get_document_id(row),
                                        '_index': self._index,
                                        '_type': self.DOC_TYPE})
                continue
            mget_map[self._get_document_id(row)] = row

        if bulk_delete_ops:
            errors = self._bulk_delete(bulk_delete_ops)
        if not mget_map:
            self._check_errors(errors)
            return

        self.logger.debug("multiple get map: %s" % repr(mget_map))
        stale_rows, mget_errors = self._get_stale_rows(mget_map)
        errors += mget_errors
        update_ops = [self._create_index_op(doc_id, row, internal_client)
                      for doc_id, row in stale_rows]
        _, update_failures = elasticsearch.helpers.bulk(
            self._es_conn,
            update_ops,
            raise_on_error=False,
            raise_on_exception=False
        )
        self.logger.debug("Index operations: %s" % repr(update_ops))

        for op in update_failures:
            op_info = op['index']
            if 'exception' in op_info:
                errors.append(op_info['exception'])
            else:
                errors.append("%s: %s" % (
                    op_info['_id'], self._extract_error(op_info)))
        self._check_errors(errors)

    def _check_errors(self, errors):
        if not errors:
            return

        for error in errors:
            self.logger.error(str(error))
        raise RuntimeError('Failed to process some entries')

    def _bulk_delete(self, ops):
        errors = []
        success_count, delete_failures = elasticsearch.helpers.bulk(
            self._es_conn, ops,
            raise_on_error=False,
            raise_on_exception=False
        )

        for op in delete_failures:
            op_info = op['delete']
            if op_info['status'] == 404:
                if op_info.get('result') == 'not_found':
                    continue
                # < 5.x Elasticsearch versions do not return "result"
                if op_info.get('found') is False:
                    continue
            if 'exception' in op_info:
                errors.append(op_info['exception'])
            else:
                errors.append("%s: %s" % (op_info['_id'],
                                          self._extract_error(op_info)))
        return errors

    def _get_stale_rows(self, mget_map):
        errors = []
        stale_rows = []
        results = self._es_conn.mget(body={'ids': mget_map.keys()},
                                     index=self._index,
                                     refresh=True,
                                     _source=['x-timestamp'])
        docs = results['docs']
        for doc in docs:
            row = mget_map.get(doc['_id'])
            if not row:
                errors.append("Unknown row for ID %s" % doc['_id'])
                continue
            if 'error' in doc:
                errors.append("Failed to query %s: %s" % (
                              doc['_id'], str(doc['error'])))
                continue
            object_date = self._get_last_modified_date(row)
            # ElasticSearch only supports milliseconds
            object_ts = int(float(object_date) * 1000)
            if not doc['found'] or object_ts > doc['_source'].get(
                    'x-timestamp', 0):
                stale_rows.append((doc['_id'], row))
                continue
        self.logger.debug("Stale rows: %s" % repr(stale_rows))
        return stale_rows, errors

    def _create_index_op(self, doc_id, row, internal_client):
        swift_hdrs = {'X-Newest': True}
        meta = internal_client.get_object_metadata(
            self._account, self._container, row['name'], headers=swift_hdrs)
        op = {'_op_type': 'index',
              '_index': self._index,
              '_type': self.DOC_TYPE,
              '_source': self._create_es_doc(meta, self._account,
                                             self._container,
                                             row['name'].decode('utf-8'),
                                             self._parse_json),
              '_id': doc_id}
        if self._pipeline:
            op['pipeline'] = self._pipeline
        return op

    """
        Verify document mapping for the elastic search index. Does not include
        any user-defined fields.
    """
    def _verify_mapping(self):
        index_client = elasticsearch.client.IndicesClient(self._es_conn)
        try:
            mapping = index_client.get_mapping(index=self._index,
                                               doc_type=self.DOC_TYPE)
        except elasticsearch.TransportError as e:
            if e.status_code != 404:
                raise
            if e.error != 'type_missing_exception':
                raise
            mapping = {}
        if not mapping.get(self._index, None) or \
                self.DOC_TYPE not in mapping[self._index]['mappings']:
            missing_fields = self.DOC_MAPPING.keys()
        else:
            current_mapping = mapping[self._index]['mappings'][
                self.DOC_TYPE]['properties']
            # We are not going to force re-indexing, so won't be checking the
            # mapping format
            missing_fields = [key for key in self.DOC_MAPPING.keys()
                              if key not in current_mapping]
        if missing_fields:
            new_mapping = dict([(k, v) for k, v in self.DOC_MAPPING.items()
                                if k in missing_fields])
            # Elasticsearch 5.x deprecated the "string" type. We convert the
            # string fields into the appropriate 5.x types.
            # TODO: Once we remove  support for the 2.x clusters, we should
            # remove this code and create the new mappings for each field.
            if self._server_version >= StrictVersion('5.0'):
                new_mapping = dict([(k, self._update_string_mapping(v))
                                    for k, v in new_mapping.items()])
            index_client.put_mapping(index=self._index, doc_type=self.DOC_TYPE,
                                     body={'properties': new_mapping})

    @staticmethod
    def _create_es_doc(meta, account, container, key, parse_json=False):
        def _parse_document(value):
            try:
                return json.loads(value.decode('utf-8'))
            except ValueError:
                return value.decode('utf-8')

        es_doc = {}
        # ElasticSearch only supports millisecond resolution
        es_doc['x-timestamp'] = int(float(meta['x-timestamp']) * 1000)
        # Convert Last-Modified header into a millis since epoch date
        ts = email.utils.mktime_tz(
            email.utils.parsedate_tz(meta['last-modified'])) * 1000
        es_doc['last-modified'] = ts
        es_doc['x-swift-object'] = key
        es_doc['x-swift-account'] = account
        es_doc['x-swift-container'] = container

        user_meta_keys = dict(
            [(k.split(MetadataSync.USER_META_PREFIX, 1)[1].decode('utf-8'),
              _parse_document(v) if parse_json else v.decode('utf-8'))
             for k, v in meta.items()
             if k.startswith(MetadataSync.USER_META_PREFIX)])
        es_doc.update(user_meta_keys)
        for field in MetadataSync.DOC_MAPPING.keys():
            if field in es_doc:
                continue
            if field not in meta:
                continue
            es_doc[field] = meta[field]
        return es_doc

    @staticmethod
    def _get_last_modified_date(row):
        ts, content, meta = decode_timestamps(row['created_at'])
        # NOTE: the meta timestamp will always be latest, as it will be updated
        # when content type is updated
        return meta

    @staticmethod
    def _extract_error(err_info):
        if 'error' not in err_info or 'root_cause' not in err_info['error']:
            return str(err_info['status'])

        err = err_info['error']['root_cause']
        try:
            return '%s: %s' % (err, err_info['error']['caused_by']['reason'])
        except KeyError:
            return err

    @staticmethod
    def _update_string_mapping(mapping):
        if mapping['type'] != 'string':
            return mapping
        if 'index' in mapping and mapping['index'] == 'not_analyzed':
            return {'type': 'keyword'}
        # This creates a mapping that is both searchable as a text and keyword
        # (the default  behavior in Elasticsearch for 2.x string types).
        return {
            'type': 'text',
            'fields': {
                'keyword': {
                    'type': 'keyword'}
            }
        }

    def _get_document_id(self, row):
        return hashlib.sha256(
            '/'.join([self._account.encode('utf-8'),
                      self._container.encode('utf-8'),
                      row['name']])
        ).hexdigest()
